File: LC/java/Valid Number.java
public class Solution {
    public boolean isNumber(String s) {
        int i = 0, e = s.length()-1;
        
        //delete the space around the string
        while(i<=e && Character.isWhitespace(s.charAt(i))){ 
            i++;  
        }
        if (i>e){ 
            return false; 
        }
        while(e>=i && Character.isWhitespace(s.charAt(e))){ 
            e--;
        }
        
        // ignore the sign
        if(s.charAt(i)=='+' || s.charAt(i)=='-'){ i++; }
        
        boolean num = false;
        boolean dot = false;
        boolean exp = false;
        
        while(i<=e){
            char curr = s.charAt(i);
            if(Character.isDigit(curr)){
                num = true;
            }
            else if(curr == '.'){
                if(dot || exp){
                    return false;
                }
                dot = true;
            }
            else if(curr == 'e'){
                if(exp || num == false){
                    return false;
                }
                exp = true;
                num = false;
            }
            else if(curr == '+' || curr == '-'){
                if(s.charAt(i-1)!='e'){
                    return false;
                }
            }
            else{
                return false;
            }
            i++;
        }
        return num;
    }
}

File: LC/python/Valid Number.py
class Solution:
    # @param s, a string
    # @return a boolean
    def isNumber(self, s):
        s=s.strip()
        
        i = 0
        e = len(s)-1
        
        if i>e:
            return False
        
        elif s[0] == '+' or s[0] == '-':
            i+=1
        
        num=False
        dot=False
        exp=False
        
        while i<=e:
            temp = s[i]
            if temp.isdigit():
                num = True
            elif temp == '.':
                if dot or exp:
                    return False
                dot = True
            elif temp == 'e':
                if exp or num == False:
                    return False
                exp = True
                num = False
            elif temp == '+' or temp == '-':
                if s[i-1]!='e':
                    return False
            else:
                return False
            i+=1
        return num
